FreeRTOS/FreeRTOS-Plus-TCP
# Adding mDNS and more changes to DNS lookups.

Description
-----------
DNS Changes:

● DNS servers can now also be reached by their IPv6 address.

● The types may be mixed: lookup a "AAAA" host on an IPv4 DNS server, but also lookup an "A" host on a DNS with a IPv6 address.

● The application hook `xApplicationDNSQueryHook()` can now set the IP-address that is to be returned. A new bit-field is introduced for this: `pxEndPoint->usDNSType, which will either be `dnsTYPE_A_HOST`, `dnsTYPE_AAAA_HOST`, or `dnsTYPE_ANY_HOST`.

● When `ipconfigUSE_MDNS` is defined, the device will respond to mDNS messages looking for "xxx.local". This will also use `xApplicationDNSQueryHook()`.

● Stop using `prvSkipNameField()`, only use `prvReadNameField()`.

● Better parsing of DNS replies: always follow a 0xC0 redirection ( see `prvReadNameField()` ).


Test Steps
-----------
I used my testing application on both a Zynq and an STM32F474. I had a Raspberry PI serving as an IPv6 router to the Internet.
I gave names to my device like `STMv4.local` and `STMv6.local`. The mDNS answer would either return an IPv4, or an IPv6 IP-address, depending on the name looked up.

The device is also able to lookup other devices that have a mDNS name ending with `.local`.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
